### ISSUE TITLE
Tests fail when using coverage flag and `getSelectorStub`

### ISSUE THREAD

**Issue by HarelM** [angular-redux2/store#9]
### This is a...

* [x] bug report

### What toolchain are you using for transpilation/bundling?

* [x] @angular/cli

### Environment

NodeJS Version: 16.13
Typescript Version: 4.5.4
Angular Version: 13.1.2
@angular-redux/store version: 2.1.3
@angular/cli version: (if applicable)  13.1.3
OS: mac and ubuntu

### Link to repo showing the issue
(optional, but helps _a lot_)
https://github.com/IsraelHikingMap/Site/tree/main/IsraelHiking.Web

### Expected Behaviour:
Test should pass allways

### Actual Behaviour:
When using coverage report and mocking a `@select` in the test the test fail since it doesn't find the right selector.
When I debug this I see the following problem:
The key is: `"(state) => state.routes.present"` but it can't be found since the coverage is wrapping this functionality changing the key name (I'm using functions to avoid incorrect spelling of hardcoded strings).
<img width="649" alt="image" src="https://user-images.githubusercontent.com/3269297/160908825-898e7db5-bbad-47d2-90be-25a190723fe3.png">

The following is my selector:
https://github.com/IsraelHikingMap/Site/blob/a629f410f543739084396a0431247ba0d06d1359/IsraelHiking.Web/src/application/services/layers/routelayers/selected-route.service.ts#L41
The following is the test that fails when using coverage:
https://github.com/IsraelHikingMap/Site/blob/a629f410f543739084396a0431247ba0d06d1359/IsraelHiking.Web/src/application/services/layers/routelayers/selected-route-service.spec.ts#L41

### Stack Trace/Error Message:
No error, just the test fails

### Additional Notes:
(optional)
Also worth noting that when using the same selector in the test but without `()` i.e.
```const routesStub: Subject<RouteData[]> = MockNgRedux.getSelectorStub<ApplicationState, RouteData[]>(
                state => state.routes.present // note here - no () around state
            );
```
The test also fails, which means this `getSelectorStub` is very fragile...

I'll see if I can somehow improve the robustness of `getSelectorStub`, but I can't promise much, any help would be appreciated...

**Comment by HarelM** [comment]
As an initial hack I managed to do the following in order to make the test pass:
```
let selectorKey = Object.keys(MockNgRedux.getSubStore().selections).find(k => k.includes("state.routes.present"));
const routesStub: Subject<RouteData[]> = MockNgRedux.getSelectorStub<ApplicationState, RouteData[]>(selectorKey);
```
Instead of 
```
const routesStub: Subject<RouteData[]> = MockNgRedux.getSelectorStub<ApplicationState, RouteData[]>(
                (state) => state.routes.present
            );
```
It does work but it's problematic when it comes to refactoring obviously...

I've also tried to use the comparator, but it is ignored at this point in time so it's not the right direction I believe.

**Comment by HarelM** [comment]
Any updates on this?
I'm currently using the following utility, but it would be better to have this somehow inside this lib, right?
```ts
const getSubject = <T>(predecator: (state: ApplicationState) => T): Subject<T> => {
    let predecatorString = predecator.toString().split("=>")[1];
    let selectorKey = Object.keys(MockNgRedux.getSubStore().selections).find(k => k.includes(predecatorString));
    return MockNgRedux.getSelectorStub<ApplicationState, T>(selectorKey);
};
```

**Comment by Garefild** [comment]
I've been a bit busy, I'll take care of it at the end of the week.
And finish all upgrades.

**Comment by HarelM** [comment]
Any updates on this? I saw you released a new version with good stuff :-)

**Comment by Garefild** [comment]
Wait a little longer and I will release another version with improvements.

**Comment by HarelM** [comment]
Will this be solved as part of next release by any chance? Let me know if there's anything I can do to help.

**Comment by Garefild** [comment]
yes

**Comment by HarelM** [comment]
Cool thanks!!!

**Comment by Garefild** [comment]
I finish all improvements to the standard lib and 154 unit test.
just left me to make the mock more stable and reliable.

I will release a beta version you can test in a few days.

**Comment by HarelM** [comment]
Great! Ping me when there's a version available with some basic instructions and I'll be happy to try it out.

**Comment by Garefild** [comment]
How can I run your project? 
To be sure that I solved the bug.

**Comment by HarelM** [comment]
clone, 
```
cd IsraelHiking.Web
npm i
npm run test -- --code-coverage --watch=false
```

Remove the following method:
https://github.com/IsraelHikingMap/Site/blob/36a7c61feef2133b295205099711c5d303c8604a/IsraelHiking.Web/src/application/services/selected-route-service.spec.ts#L12
And replace it with the relevant changes you made.

Thanks!

**Comment by Garefild** [comment]
Hi, I am working on this fast as I can.
End of the year and my work pressure on me :( 
Sorry about the delays, night and weekend this is what I have.

a few points that I don't understand.
and improvements to me to understand so I can make developer life easier when they write the tests.

** You can copy and paste and it will work with the old version of the lib (I want to investigate the bug further) **
> I did install the new lib FYI


I would appreciate it if you could check and let me know if it works for you too.
This will greatly affect library changes and upgrades.
I have already thought of some improvements to the code to help tempts and make the tests easier.

Thanks!

1. Why do you use inject?

this is a more readable way:
```typescript
    let service: SelectedRouteService;

    beforeEach(() => {
        let toastMock = new ToastServiceMockCreator();
        toastMock.resourcesService.route = "route";
        let routerServiceMock = {};
        TestBed.configureTestingModule(...));

        MockNgRedux.reset();
        service = TestBed.inject(SelectedRouteService);
    });

// -----

    it("Should get selected route when there are routes", () => {
            const routesStub = getSubject((state: ApplicationState) => state.routes.present);

            routesStub.next([{} as any]);

            let selectedRoute = service.getSelectedRoute();

            expect(selectedRoute).not.toBeUndefined();
        }
    );

```

2. This is the selector you are wont to test "(state: ApplicationState) => state.routeEditingState.selectedRouteId)"
but you are using (state: ApplicationState) => state.routes.present) why?
> "getSelectedRoute" => "this.selectedRouteId" that is set from subscribe of "this.selectedRouteId$"

3. why did you not test the subscription in isolation something like this?: 
```typescript 
    it("Should get selected route when there are routes", () => {
        const routesStub = MockNgRedux.getSelectorStub(
            (state: ApplicationState) => state.routeEditingState.selectedRouteId
        );

        routesStub.next("someID");
        routesStub.complete();

        (service as any).selectedRouteId$.subscribe((id: string) => {
            expect(id).toBe("someID");
        });
    });

// this work for me as well.

    it("Should get selected route when there are routes", () => {
        const routesStub = MockNgRedux.getSelectorStub(
            (state: ApplicationState) => state.routeEditingState.selectedRouteId
        );

        routesStub.next("someID");
        routesStub.complete();

        (service as any).selectedRouteId$.subscribe((id: string) => {
            let selectedRoute = service.getSelectedRoute();
            expect(selectedRoute).not.toBeUndefined();
            expect(selectedRoute.name).toBe("SomeID");
        });
    });

```

**Comment by HarelM** [comment]
>Why do you use inject?

A habit I guess, this way I can easily add more services to the test without needing to initialize them up front and I know which service each test it using. nothing dramatic here...

>This is the selector you are wont to test ...

Not sure I understand the question here, I'm basically testing what happens when there no routes and I request the selected route. It might be that I have changed the logic and didn't change the test though... I'll look into it.

>why did you not test the subscription in isolation something like this?

If I'm doing ` (service as any)` it means I'm not testing the public API, and for me that's a big no-no... I should test the public API or the things that the service is doing when calling other services, and not an internal field or method.

**Comment by HarelM** [comment]
Thanks for all the work! I truly appreciate it, and I hope that you are not getting the wrong impression from my comments.
As a side note, I don't see any branches in this repo so you are probably working on this locally only.
I might be able to help if you publish an initial branch with your work.

**Comment by Garefild** [comment]
My problem is that I can't fully understand the bug and in cent restore it.
That's why I didn't make any significant changes,
just debug your code with an old version of the lib.
I will create a new branch with the change.

> Not sure I understand the question here, I'm testing what happens when there are no routes and I request the selected route. It might be that I have changed the logic and didn't change the test though... I'll look into it.

https://github.com/IsraelHikingMap/Site/blob/36a7c61feef2133b295205099711c5d303c8604a/IsraelHiking.Web/src/application/services/selected-route-service.spec.ts#L46

```typescript
// this is the original test 
    it("Should get selected route when there are routes", inject([SelectedRouteService],
        (selectedRouteService: SelectedRouteService) => {
        // this is the wrong selector u spouse to use ** (state: ApplicationState) => state.routeEditingState.selectedRouteId) **
            const routesStub = getSubject((state: ApplicationState) => state.routes.present);

            routesStub.next([{} as any]);

// this happens before the async operation so u get undefined.
// the function getSubject add a delay so this is why it works.
            let selectedRoute = selectedRouteService.getSelectedRoute();

            expect(selectedRoute).not.toBeUndefined();
        }
    ));
```

if you replace the code with this on your system it will work:

```typescript
    it("Should get selected route when there are routes", inject([ SelectedRouteService ],
        (selectedRouteService: SelectedRouteService) => {
            const routesStub = MockNgRedux.getSelectorStub((state: ApplicationState) => state.routeEditingState.selectedRouteId));

            routesStub.next([ {} as any ]);

            setTimeout(() => {
                let selectedRoute = selectedRouteService.getSelectedRoute();
                expect(selectedRoute).not.toBeUndefined();
            }, 1000);
        }
    ));
```

**Comment by Garefild** [comment]
Do u have any chat that we can talk?

**Comment by HarelM** [comment]
Facebook messanger? https://www.facebook.com/harel.mazor/

**Comment by HarelM** [comment]
The above test is a mistake, as you said, it is an old left over from a code I removed, and the test is written in a way that still passes even though I removed the relevant logic due to missing `id` field in the routes.

Can you check if you add the following code in the above test, does it pass for you:
```js
it("Should check getSelectorStub", inject([SelectedRouteService],
        (selectedRouteService: SelectedRouteService) => {
            const selectedRouteIdSubject = MockNgRedux.getSelectorStub((state: ApplicationState) => state.routeEditingState.selectedRouteId);
            selectedRouteIdSubject.next("1");

            expect((selectedRouteService as any).selectedRouteId).toBeDefined();
        }
));
```
I've also tested that with a timeout (remember that you need to wait for the timeout to execute otherwise the test will pass simply because the timeout didn't ran).

I've updated the main branch with some minor changes there.

**Comment by Garefild** [comment]
![image](https://user-images.githubusercontent.com/2126092/204906508-5af4452e-9fc2-498b-8ef5-8a89b14fe782.png)

```typescript
    it("Should check getSelectorStub", inject([ SelectedRouteService ],
        (selectedRouteService: SelectedRouteService) => {
            // Create Subject for selector.
            const selectedRouteIdSubject = MockNgRedux.getSelectorStub(
                (state: ApplicationState) => state.routeEditingState.selectedRouteId
            );

            // emit 1 to subject that create for selector.
            selectedRouteIdSubject.next("1");

            // run before selectedRouteId in update from the change.
            // expect((selectedRouteService as any).selectedRouteId).toBeDefined();

            // create delay know that subject emit the async msg.
            selectedRouteIdSubject.subscribe(() => {
                console.log("print msg to validate the expect work..");
                expect((selectedRouteService as any).selectedRouteId).toBeDefined();
            });
        }
    ));
```

**Comment by HarelM** [comment]
Thanks, this works, but that's not the original issue. :-)
Please use the above test and run it with coverage:
`npm run test -- --code-coverage --watch=false`
![image](https://user-images.githubusercontent.com/3269297/204908492-39c7f2e5-7286-466d-9515-1309f9227cb3.png)

**Comment by Garefild** [comment]
Ummm, let's see what we dealing with.
I'm running some tests now, I must recreate the bug on my side.

**Comment by Garefild** [comment]
You can run the next code and tell me what you get. 

```typescript
    it("Should check getSelectorStub", inject([ SelectedRouteService ],
        (selectedRouteService: SelectedRouteService) => {
            const selectedRouteIdSubject = MockNgRedux.getSelectorStub(
                (state: ApplicationState) => state.routeEditingState.selectedRouteId
            );
            selectedRouteIdSubject.next("1");

            selectedRouteIdSubject.subscribe((id) => {
                expect(id).toBe("1");
            });

            // expect((selectedRouteService as any).selectedRouteId).toBeDefined();
        }
    ));
```

**Comment by Garefild** [comment]
![image](https://user-images.githubusercontent.com/2126092/204915154-9f3b50a6-afa4-407b-b7e5-741ad0f6f73b.png)

**Comment by Garefild** [comment]
I found the little bastard !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by HarelM** [comment]
Great to hear! note that in my initial post I did some debugging as to the root cause of this issue, which is caused by the instrumentation of the code coverage. But that doesn't help to define the solution though...

**Comment by Garefild** [comment]
Long story short this is the bug: 

```js
Object{state => {
  cov_jv20rt6qg().f[5]++;
  cov_jv20rt6qg().s[9]++;
  return state.routeEditingState.selectedRouteId;
}
```

I'm working on a fix ...

**Comment by HarelM** [comment]
Great!

**Comment by Garefild** [comment]
OK, the bug is fixed.
It happened because "istanbuljs" library changed the function on run-time to detect coverage.
The changes make the detection of anonymous functions impossible.
I need a better way to detect the functions, I'm redesigning the selection part at the weekend.

at the same spirit why are you using a function selection? 
it is for logic selection and not for static selection. 

```typescript
    @Select((state: ApplicationState) => state.routes.present)
    private routes$: Observable<RouteData[]>;

    @Select((state: ApplicationState) => state.routeEditingState.selectedRouteId)
    private selectedRouteId$: Observable<string>;
```

=> 

```typescript
    @Select(['routes', 'present'])
    private routes$: Observable<RouteData[]>;

    @Select(['routeEditingState', 'selectedRouteId'])
    private selectedRouteId$: Observable<string>;
```

**Comment by Garefild** [comment]
the fixed code (not the best way but works). 

```typescript
   protected initSelectorStubFunction<SelectedState>(
        key: string,
        selector: FunctionSelector<any, any>
    ): string {
        let functionID = 'generic_function';
        const functionReturn = key.match(/return.+?;|=>.+(?!{)./);

        if (functionReturn) {
            functionID = functionReturn[0];

            if (functionID.includes('=>')) {
                functionID = functionID.replace('=>', 'return') + ';';
            }
        }

        try {
            selector({});
        } catch (e) { /* empty */ }

        return functionID;
    }
}
```

**Comment by HarelM** [comment]
Reminds the fix I did...
I'm using function selection to allow refactoring, find usages etc... Type safety basically...

**Comment by Garefild** [comment]
The problem with this type of fix is that it allows for duplication.
Especially if there is logic in the function like the original purpose of this selector.
Therefore I need to design a much more rigid solution.
Because the solution I wrote above has a lot of freedom to override similar selectors.

**Comment by HarelM** [comment]
Yea... That's what I thought too, but maybe you can add this as a flag - i.e. "loose finding" for this specific case.

**Comment by Garefild** [comment]
OK, I designed a more robust solution. 
now I am left with endless time to recreate comments && documentation && tutorials && example project.

**Comment by Garefild** [comment]
I modify your angular web to the new design to check for end cases in the real world :)
So far the redux work is not too shabby XD

I didn't dive into your code I change only what I must. 
in the new version, there is a lot of optimization that you can use to reduce your code.

before: 
https://github.com/IsraelHikingMap/Site/blob/36a7c61feef2133b295205099711c5d303c8604a/IsraelHiking.Web/src/application/reducers/layers.reducer.ts

after: 
```typescript
import { orderBy, remove, some } from "lodash-es";
import { AbstractReducer, Action, ActionPayload, AnyAction } from "@angular-redux2/store";
import type { LayersState, EditableLayer, Overlay, CategoriesGroupType, Category } from "../models/models";
import { initialState, ISRAEL_HIKING_MAP, ISRAEL_MTB_MAP, SATELLITE, HIKING_TRAILS, BICYCLE_TRAILS } from "./initial-state";

export type AddOverlayPayload = { layerData: Overlay };
export type ToggleGroupPayload = { name: string};
export type RemoveLayerPayload = { key: string };
export type AddBaseLayerPayload = { layerData: EditableLayer };
export type SelectBaseLayerPayload = {key: string};

export type UpdateBaseLayerPayload = {
    key: string;
    layerData: EditableLayer;
};

export type UpdateOverlayPayload = {
    key: string;
    layerData: Overlay;
};

export type SetCategoriesGroupVisibilityPayload = {
    groupType: CategoriesGroupType;
    visible: boolean;
};

export type AddCategoryPayload = {
    groupType: CategoriesGroupType;
    category: Category;
};

export type UpdateCategoryPayload = {
    groupType: CategoriesGroupType;
    category: Category;
};

export type RemoveCategoryPayload = {
    groupType: CategoriesGroupType;
    categoryName: string;
};

export type SetCategoryVisibilityPayload = {
    name: string;
    groupType: CategoriesGroupType;
    visible: boolean;
};

export type ToggleOfflinePayload = {
    key: string;
    isOverlay: boolean;
};

export class LayersReducer extends AbstractReducer {
    static actions: {
        addOverlay: ActionPayload<AddOverlayPayload>;
        expandGroup: ActionPayload<ToggleGroupPayload>;
        addCategory: ActionPayload<AddCategoryPayload>;
        LayersState: ActionPayload<ToggleOfflinePayload>;
        addBaseLayer: ActionPayload<AddBaseLayerPayload>;
        collapseGroup: ActionPayload<ToggleGroupPayload>;
        updateOverlay: ActionPayload<UpdateOverlayPayload>;
        removeOverlay: ActionPayload<RemoveLayerPayload>;
        updateCategory: ActionPayload<UpdateCategoryPayload>;
        removeCategory: ActionPayload<RemoveCategoryPayload>;
        removeBaseLayer: ActionPayload<RemoveLayerPayload>;
        selectBaseLayer: ActionPayload<SelectBaseLayerPayload>;
        updateBaseLayer: ActionPayload<UpdateBaseLayerPayload>;
        setCatgegoryVisibility: ActionPayload<SetCategoryVisibilityPayload>;
        setCatgegoriesGroupVisibility: ActionPayload<SetCategoriesGroupVisibilityPayload>;
    };

    private sort(layers: EditableLayer[]): EditableLayer[] {
        let ordered = orderBy(layers, l => l.key);
        let specialKeys = [ ISRAEL_HIKING_MAP, ISRAEL_MTB_MAP, SATELLITE, HIKING_TRAILS, BICYCLE_TRAILS ];
        let removed = remove(ordered, o => specialKeys.indexOf(o.key) !== -1);
        ordered = [ ...removed, ...ordered ];
        return ordered;
    }

    @Action
    public addBaseLayer(lastState: LayersState, action: AnyAction<AddBaseLayerPayload>): void {
        lastState.baseLayers = this.sort([ ...lastState.baseLayers, action.payload.layerData ]);
    }

    @Action
    public addOverlay(lastState: LayersState, action: AnyAction<AddOverlayPayload>): LayersState {
        lastState.overlays = this.sort([ ...lastState.overlays, action.payload.layerData ]) as Overlay[];

        return lastState;
    }

    @Action
    public removeBaseLayer(lastState: LayersState, action: AnyAction<RemoveLayerPayload>): void {
        let baseLayers = lastState.baseLayers;
        baseLayers.splice(baseLayers.indexOf(baseLayers.find(b => b.key === action.payload.key)), 1);
    }

    @Action
    public removeOverlay(lastState: LayersState, action: AnyAction<RemoveLayerPayload>): LayersState {
        let overlays = lastState.overlays;
        overlays.splice(overlays.indexOf(overlays.find(o => o.key === action.payload.key)), 1);

        return lastState;
    }

    @Action
    public updateBaseLayer(lastState: LayersState, action: AnyAction<UpdateBaseLayerPayload>): LayersState {
        let baseLayers = lastState.baseLayers;
        baseLayers.splice(baseLayers.indexOf(
            baseLayers.find(b => b.key === action.payload.key)
        ), 1, action.payload.layerData);
        lastState.baseLayers = this.sort(baseLayers);

        return lastState;
    }

    @Action
    public updateOverlay(lastState: LayersState, action: AnyAction<UpdateOverlayPayload>): void {
        let overlays = lastState.overlays;
        overlays.splice(overlays.indexOf(
            overlays.find(o => o.key === action.payload.key)
        ), 1, action.payload.layerData);
        lastState.overlays = this.sort(overlays) as Overlay[];
    }

    @Action
    public selectBaseLayer(lastState: LayersState, action: AnyAction<SelectBaseLayerPayload>): void {
        lastState.selectedBaseLayerKey = action.payload.key;
    }

    @Action
    public expandGroup(lastState: LayersState, action: AnyAction<ToggleGroupPayload>): LayersState {
        let expanded = lastState.expanded;
        if (expanded.find(n => n === action.payload.name) != null) {
            return lastState;
        }
        expanded.push(action.payload.name);

        return lastState;
    }

    @Action
    public collapseGroup(lastState: LayersState, action: AnyAction<ToggleGroupPayload>): LayersState {
        let expanded = lastState.expanded;
        if (expanded.find(n => n === action.payload.name) == null) {
            return lastState;
        }
        expanded.splice(expanded.indexOf(action.payload.name));

        return lastState;
    }

    @Action
    public addCategory(lastState: LayersState, action: AnyAction<AddCategoryPayload>): LayersState {
        let groups = lastState.categoriesGroups;
        let group = groups.find(g => g.type === action.payload.groupType);
        let categories = group.categories;
        categories.push(action.payload.category);
        let newGroup = {
            ...group,
            categories
        };
        groups.splice(groups.indexOf(group), 1, newGroup);

        lastState.categoriesGroups = groups;

        return lastState;
    }

    @Action
    public updateCategory(lastState: LayersState, action: AnyAction<UpdateCategoryPayload>): void {
        let groups = lastState.categoriesGroups;
        let group = groups.find(g => g.type === action.payload.groupType);
        let categories = [ ...group.categories ];
        let categoryIndex = categories.indexOf(categories.find(c => c.name === action.payload.category.name));
        categories.splice(categoryIndex, 1, action.payload.category);
        let newGroup = {
            ...group,
            categories
        };
        groups.splice(groups.indexOf(group), 1, newGroup);
        lastState.categoriesGroups = groups;
    }

    @Action
    public removeCategory(lastState: LayersState, action: AnyAction<RemoveCategoryPayload>): LayersState {
        let groups = lastState.categoriesGroups;
        let group = groups.find(g => g.type === action.payload.groupType);
        let categories = [ ...group.categories ];
        let categoryIndex = categories.indexOf(categories.find(c => c.name === action.payload.categoryName));
        categories.splice(categoryIndex, 1);
        let newGroup = {
            ...group,
            categories
        };
        groups.splice(groups.indexOf(group), 1, newGroup);
        lastState.categoriesGroups = groups;

        return lastState;
    }

    @Action
    public setCatgegoryVisibility(lastState: LayersState, action: AnyAction<SetCategoryVisibilityPayload>): void {
        let groups = lastState.categoriesGroups;
        let group = groups.find(g => g.type === action.payload.groupType);
        let categories = [ ...group.categories ];
        let category = categories.find(c => c.name === action.payload.name);
        let newCategory = {
            ...category,
            visible: action.payload.visible
        };
        categories.splice(categories.indexOf(category), 1, newCategory);
        let newGroup = {
            ...group,
            categories,
            visible: some(categories, c => c.visible)
        };
        groups.splice(groups.indexOf(group), 1, newGroup);
        lastState.categoriesGroups = groups;
    }

    @Action
    public setCatgegoriesGroupVisibility(lastState: LayersState, action: AnyAction<SetCategoriesGroupVisibilityPayload>): void {
        let groups = lastState.categoriesGroups;
        let group = groups.find(g => g.type === action.payload.groupType);
        let categories = [];
        for (let category of group.categories) {
            categories.push({
                ...category,
                visible: action.payload.visible
            });
        }
        let newGroup = {
            ...group,
            categories,
            visible: action.payload.visible
        };
        groups.splice(groups.indexOf(group), 1, newGroup);
        lastState.categoriesGroups = groups;
    }

    @Action
    public toggleOffline(lastState: LayersState, action: AnyAction<ToggleOfflinePayload>): void {
        if (action.payload.isOverlay) {
            let overlays = lastState.overlays;
            let layer = overlays.find(b => b.key === action.payload.key);

            let layerData = {
                ...layer,
                isOfflineOn: !layer.isOfflineOn
            } as Overlay;
            overlays.splice(overlays.indexOf(layer), 1, layerData);
            lastState.overlays = overlays;

            return;
        } else {
            let baseLayers = lastState.baseLayers;
            let layer = baseLayers.find(b => b.key === action.payload.key);
            let layerData = {
                ...layer,
                isOfflineOn: !layer.isOfflineOn
            } as EditableLayer;
            baseLayers.splice(baseLayers.indexOf(layer), 1, layerData);
        }
    }
}

export const layersReducer = LayersReducer.createReducer(initialState.layersState);

```

**Comment by Garefild** [comment]
![image](https://user-images.githubusercontent.com/2126092/205377485-69c592f0-1418-46ea-b2ee-096597cf9a50.png)

**Comment by Garefild** [comment]
```typescript
import { Component, Input } from "@angular/core";
import { Dispatch, NgRedux } from "@angular-redux2/store";

import { BaseMapComponent } from "../base-map.component";
import { ResourcesService } from "../../services/resources.service";
import {
    LayersReducer,
} from "../../reducers/layers.reducer";
import type { ApplicationState, CategoriesGroup, Category } from "../../models/models";

@Component({
    selector: "categories-group",
    templateUrl: "./categories-group.component.html"
})
export class CategoriesGroupComponent extends BaseMapComponent {

    @Input()
    public categoriesGroup: CategoriesGroup;

    constructor(resources: ResourcesService,
                private readonly ngRedux: NgRedux<ApplicationState>) {
        super(resources);
    }

    @Dispatch
    public expand() {
        return LayersReducer.actions.expandGroup({
            name: this.categoriesGroup.type
        });
    }

    @Dispatch
    public collapse() {
        return LayersReducer.actions.collapseGroup({
            name: this.categoriesGroup.type
        });
    }

    public getExpandState(): boolean {
        return this.ngRedux.getState().layersState.expanded.find(l => l === this.categoriesGroup.type) != null;
    }

    public toggleCategory(category: Category) {
        this.ngRedux.dispatch(
            LayersReducer.actions.setCatgegoryVisibility({
                groupType: this.categoriesGroup.type,
                name: category.name,
                visible: !category.visible
            })
        );
    }

    @Dispatch
    public toggleVisibility(event: Event) {
        event.stopPropagation();

        return LayersReducer.actions.setCatgegoriesGroupVisibility({
            groupType: this.categoriesGroup.type,
            visible: !this.categoriesGroup.visible
        });
    }
}
```

**Comment by Garefild** [comment]
![image](https://user-images.githubusercontent.com/2126092/205379582-dd97d18e-8294-40cd-9e6b-f64837d7a43e.png)

**Comment by HarelM** [comment]
Interesting :-) automatically generating the actions from the class public methods would be amazing as I think you suggested, but this is also great! Can't wait for the new version!

**Comment by Garefild** [comment]
I had issues with the autocomplete in the old way so I had to think in a different way and this is the result.
this is for now, and I will find a way to generate all automatically.

**Comment by Garefild** [comment]
Hi, 5.0.0-beta.1 has publish :)
https://www.npmjs.com/package/@angular-redux2/store/v/5.0.0-beta.1

**Comment by HarelM** [comment]
I've upgraded to the beta and the test pass on both regular and code coverage stuff.
Thanks for the fix!!

**Comment by Garefild** [comment]
Are there things to improve? 
Was the transition okay?

**Comment by HarelM** [comment]
For the testing stuff this was simple and straight forward.
I remember I had to search deeply in this library to find the docs on how to write test with the mock stuff, I would consider improving the docs and making note about how to test the `@Select(...)` stuff.
The transition from calling an Action to the "generated" action field was where most of the work was, but it makes sense as this is a major improvement and a lot of code was removed.
